File: Search/Search a 2D Matrix/Solution.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    if (matrix === null || matrix.length === 0) {
        return false;
    }
    if (matrix[0] === null || matrix[0].length === 0) {
        return false;
    }
    
    var row = matrix.length;
    var col = matrix[0].length;
    var first = 0;
    var last = row * col;
    
    while (first < last) {
        var mid = (first + last) / 2;
        var val = matrix[~~(mid / col)][~~(mid % col)];
        
        if (val === target) return true;
        else if (val < target) first = mid + 1;
        else last = mid;
    }
    
    return false;
};

File: Search/Search a 2D Matrix/Solution.cpp
// LeetCode, Search a 2D Matrix
// Time complexity O(n)
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int row = 0;
        int m = matrix.size();
        int n = matrix.front().size();
        int col = n - 1;
        
        while (row < m && col >= 0) {
            if (matrix[row][col] == target)
                return true;
            else if(matrix[row][col] < target)
                row++;
            else col--;
        }
        return false;
    }
};